src/senaite/core/supermodel/model.py

```diff
@@ -90,6 +94,25 @@ def init_with_instance(self, instance):
         self._catalog = self.get_catalog_for(instance)
         self._instance = instance
 
+    def __del__(self):
+        """Destructor
+
+        Terminates all references for garbage collection
+        """
+        logger.debug("Destroying {}".format(repr(self)))
+
+        # Re-ghostify the instance
+        # https://zodb.readthedocs.io/en/latest/api.html#persistent.interfaces.IPersistent
+        if self._instance is not None:
+            if not getattr(self._instance, "_p_changed", 0):
+                self._instance._p_deactivate()
```

Comment:
Yes, due to the check of the `_p_changed` flag no data will be thrown away.

However, regarding to the documentation, `_p_invalidate()` causes any data to be thrown away, even if the object is in the changed state, while `_p_deactivate()` possibly changes an object in the saved state to the ghost state. It may not be possible to make some persistent objects ghosts, and, for optimization reasons, the implementation may choose to keep an object in the saved state.

Therefore, I decided to go ahead with the `_p_deactivate` call.